File: javascript-leetcode/543.js
/**
 * Definition for a binary tree node.
 * function TreeNode(val) {
 *     this.val = val;
 *     this.left = this.right = null;
 * }
 */
/**
 * @param {TreeNode} root
 * @return {number}
 */
var diameterOfBinaryTree = function(root) {
  let result = 1;
  var depth = function(node) {
    if (node === null) {
      return 0;
    }
    let left = depth(node.left);
    let right = depth(node.right);
    result = Math.max(result, left + right + 1);
    return Math.max(left, right) + 1;
  };
  depth(root);
  return result - 1;
};

File: python-leetcode/543.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution:
  def diameterOfBinaryTree(self, root: TreeNode) -> int:
    self.result = 0
    def depth(root):
      if root == None: return 0
      left = depth(root.left)
      right = depth(root.right)
      self.result = max(self.result, left + right)
      return max(left, right) + 1
    depth(root)
    return self.result
